src/core/coap/coap.cpp

```diff
@@ -714,29 +714,26 @@ void CoapBase::ProcessReceivedRequest(Message &aMessage, const Ip6::MessageInfo
         break;
     }
 
-    SuccessOrExit(error = iterator.Init(&aMessage));
-    for (const otCoapOption *option = iterator.GetFirstOption(); option != nullptr; option = iterator.GetNextOption())
+    SuccessOrExit(error = iterator.Init(aMessage, kOptionUriPath));
+
+    while (!iterator.IsDone())
     {
-        switch (option->mNumber)
+        uint16_t optionLength = iterator.GetOption()->GetLength();
+
+        if (curUriPath != uriPath)
         {
-        case kOptionUriPath:
-            if (curUriPath != uriPath)
-            {
-                *curUriPath++ = '/';
-            }
+            *curUriPath++ = '/';
+        }
 
-            VerifyOrExit(option->mLength < sizeof(uriPath) - static_cast<size_t>(curUriPath + 1 - uriPath), OT_NOOP);
+        VerifyOrExit(curUriPath + optionLength < OT_ARRAY_END(uriPath), OT_NOOP);
```

Comment:
Adding  a note here about the change so that it can also be checked/verified by others/reviewers.

I think the condition in the original code was too restrictive (potentially leaving one byte/char in the array unused). In the new code, we just check if after adding `optionLength` to `cur` we are still before the array end (which should still leave space for the null `\0` char to be added).